---
title: "Metegenomics: Microbes in Hospital Air"
author: "Kohl Kinning"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook:
    fig_height: 4
    theme: cosmo
    toc: yes
    toc_float:
      collapsed: yes
---

So much interesting data is open and freely available. I'm fascinated with the biology of engineered systems. To keep myself abreast of a wide range of applications of bioinformatics tools, I dive head first in to some exploration of this metagenomic data. I'll follow the approach of (Grossart et al., 2018) in recovering secondary metabolites genes from HVAC metagenomic data produced in (King, 2016).

# Data

## Access
The data was downlaoded from the NCBI Sequence Read Archive database study number [SRP060568](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP060568) using `fastq-dump accession_list.txt`. 

## Design
From the submitters: Air samples were collected as part of a study to develop methods for monitoring air from an indoor environment using CIWGS methods by constructing NGS libraries from samples with low amounts of DNA. 

The sequencing was performed on an Illumina HiSeq 2500. The reads are single end.


*accession_list.txt*
```
SRR2183670
SRR2183684
SRR2183683
SRR2183682
SRR2183697
SRR2183696
SRR2183695
SRR2183694
SRR2183692
SRR2183691
SRR2183690
SRR2183689
SRR2183688
SRR2183687
SRR2183685
SRR2183681
SRR2183680
SRR2183679
SRR2183677
SRR2183676
SRR2183675
SRR2183674
SRR2183673
SRR2183672
SRR2183700
SRR2183706
SRR2183709
SRR2183708
SRR2183707
SRR2183705
SRR2183704
SRR2183703
SRR2183701
SRR2183699
SRR2183698
SRR2183724
SRR2183723
SRR2183722
SRR2183721
SRR2183720
SRR2183719
SRR2183717
SRR2183716
SRR2183715
SRR2183734
SRR2183735
SRR2183736
SRR2183733
SRR2183732
SRR2183731
SRR2183730
SRR2183729
SRR2183728
SRR2183727
SRR2183725
SRR2183714
SRR2183713
SRR2183712
SRR2183711
SRR2183737
SRR2183738
```

## Plan

After preparing the raw reads, I'll use MetaBat to reconstruct single genomes from these metagenomic samples. I'll then use antiSMASH to mine the genome for potentially interesting secondary metabolites in the microbial genomes. I find it fascintating that natural products which may have a use in humans can be residing in the mechanical systmes of engineered structures. Evolution of these microbes is happening at a dizzying pace due to newly introduced stable environments and massive pressure as a result of massive specation events, such as routine sanitization.

# Raw data preparation

## Download

```{bash}
#!/bin/bash
for SAMPLE in $(cat ../data/accession_list.txt); do
    prefetch $SAMPLE
done

for SAMPLE in $(cat ../data/accession_list.txt); do
    fastq-dump ~/ncbi/public/sra/$SAMPLE.sra 
done

cat accession_list.txt | awk {'print "prefetch "$1""'} > prefetch_list.sh
cat accession_list.txt | awk {'print "fastq-dump --split-files ~/ncbi/public/sra/"$1".sra"'} > convert_sra.sh

cat accession_list.txt | system(fastq-dump ~/ncbi/public/sra/$1.sra)
```


## Adapter trimming

See this [technical note](https://swiftbiosci.com/wp-content/uploads/2016/09/16-0853-Tail-Trim-TN.pdf). Remove leading and trailing 10 bases AFTER adapter trimming! Two rounds of trimmomatic.


```{r}
getwd()
```
### Trim Illumina adapters

```{bash}
wd=../data
indir=${wd}/fastq
odir=${wd}/trimmomatic
samples=$(ls $indir | grep -E ".fastq" | sed -E 's/([A-Z0-9]+)(_[0-9])(.fastq)/\1/' | uniq)

for file in $samples; do
    FASTQ1=${indir}/${file}_1.fastq
    FASTQ2=${indir}/${file}_2.fastq
    SAMPLE=$file
    ODIR=${odir}/${SAMPLE}
    mkdir ${ODIR}

    # Run trimmomatic on the sample
    java -Xmx27g -jar ~/Trimmomatic-0.38/trimmomatic-0.38.jar PE \
    -threads 28 \
    -phred33 \
    -trimlog ${ODIR}/${SAMPLE}.trimmmomatic.log \
    ${FASTQ1} ${FASTQ2} \
    ${ODIR}/${SAMPLE}_trimmed_1.fastq ${ODIR}/${SAMPLE}_trimmed_unpaired_1.fastq ${ODIR}/${SAMPLE}_trimmed_2.fastq ${ODIR}/${SAMPLE}_trimmed_unpaired_2.fastq \
    ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:8:TRUE SLIDINGWINDOW:3:20 TRAILING:3 LEADING:3 \
    1> ${ODIR}/${SAMPLE}_stats.log 2>&1
done
```

This script will perform the following: 

+ Remove Illumina adapters provided in the TruSeq3-PE.fa file (provided) 
  + initially look for seed matches (16 bases) allowing maximally 2 mismatches extend and clip seeds if a score of 30 is reached (about 50 bases), keep both farward and reverse reads
    + fastaWithAdaptersEtc: TruSeq3 adapter file
    + seed mismatches: 2 bases
    + palindrome clip threshold 30 bases
    + simple clip threshold: 10 bases
    + minAdapterLength: 8 bases
    + keepBothReads: TRUE retains the reverse read if an adapter is detected
+ Scan the read with a 3-base wide sliding window, cutting when the average quality per base drops below 20 (SLIDINGWINDOW:3:20)
+ Remove trailing low quality or N bases (below quality 3) (TRAILING:3)
+ Remove leading low quality or N bases (below quality 3) (LEADING:3)
+ Save std out (important info about read survival) in to a log file.

### Trim adaptase

```{bash}
wd=../data
indir=${wd}/trimmomatic
odir=${wd}/trimmomatic_cropped
samples=$(ls $indir)

for SAMPLE in $samples; do
    FASTQ1=${indir}/${SAMPLE}/${SAMPLE}_trimmed_1.fastq
    FASTQ2=${indir}/${SAMPLE}/${SAMPLE}_trimmed_2.fastq
    ODIR=${odir}/${SAMPLE}
    mkdir ${ODIR}

    # Run trimmomatic on the sample
    java -Xmx27g -jar ~/Trimmomatic-0.38/trimmomatic-0.38.jar PE \
    -threads 28 \
    -phred33 \
    -trimlog ${ODIR}/${SAMPLE}.trimmmomatic.log \
    ${FASTQ1} ${FASTQ2} \
    ${ODIR}/${SAMPLE}_trimmed_cropped_1.fastq ${ODIR}/${SAMPLE}_trimmed_cropped_unpaired_1.fastq ${ODIR}/${SAMPLE}_trimmed_cropped_2.fastq ${ODIR}/${SAMPLE}_trimmed_cropped_unpaired_2.fastq \
    ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 HEADCROP:10 \
    1> ${ODIR}/${SAMPLE}_stats.log 2>&1
done
```

+ Remove the leading 10 bases (HEADCROP:10) regardless of quality. This is in effert to trim out the adaptase sequences unique to this library preparation technique. See [technical note](https://swiftbiosci.com/wp-content/uploads/2016/09/16-0853-Tail-Trim-TN.pdf)

## FastQC/Mutli-QC

Perform fastQC on the trimmed data. Only continue forward with the paired reads.

 | grep -E ".fq" | sed -E 's/([A-Z0-9]+_trimmed)(_[0-9])(.fastq)/\1/' | uniq

```{bash}
wd=../data
indir=${wd}/fastq
odir=${wd}/trimmomatic_trimmed
samples=$(ls $indir/)
fastqc=$(ls $odir)

for SAMPLE in $samples; do
    if [[ ${fastqc}[*] =~ $SAMPLE ]]; then
        echo "Already generated quality report for $SAMPLE"
    else
        echo "Generating quality report for $SAMPLE"
          echo "$SAMPLE is Paired End"
          FASTQ_1=${indir}/${SAMPLE}/${SAMPLE}_trimmed_1.fastq
          FASTQ_2=${indir}/${SAMPLE}/${SAMPLE}_trimmed_2.fastq
          
          # Run FASTQC
          fastqc -o ${odir} ${FASTQ_1} -t 4 -q
          fastqc -o ${odir} ${FASTQ_2} -t 4 -q
          if [ $? -eq 0 ]
          then
              echo "Quality report for ${SAMPLE} successfully generated"
          else
              echo "Error: FASTQC failed on ${SAMPLE}" >&2
              exit 1
        fi
    fi
    fastqc=$(ls $odir)
done
```

```{bash}
wd=../data
indir=${wd}/fastq
odir=${wd}/fastqc
samples=$(ls $indir/)
fastqc=$(ls $odir)

for SAMPLE in $samples; do
    if [[ ${fastqc}[*] =~ $SAMPLE ]]; then
        echo "Already generated quality report for $SAMPLE"
    else
        echo "Generating quality report for $SAMPLE"
          echo "$SAMPLE is Paired End"
          FASTQ_1=${indir}/${SAMPLE}
          FASTQ_2=${indir}/${SAMPLE}
          
          # Run FASTQC
          fastqc -o ${odir} ${FASTQ_1} -t 4 -q
          fastqc -o ${odir} ${FASTQ_2} -t 4 -q
          if [ $? -eq 0 ]
          then
              echo "Quality report for ${SAMPLE} successfully generated"
          else
              echo "Error: FASTQC failed on ${SAMPLE}" >&2
              exit 1
        fi
    fi
    fastqc=$(ls $odir)
done
```




## Assembly

--meta
    This flag is recommended when assembling metagenomic data sets (runs metaSPAdes, see paper for more details). Currently metaSPAdes supports only a single library which has to be paired-end (we hope to remove this restriction soon). It does not support careful mode (mismatch correction is not available). In addition, you cannot specify coverage cutoff for metaSPAdes. Note that metaSPAdes might be very sensitive to presence of the technical sequences remaining in the data (most notably adapter readthroughs), please run quality control and pre-process your data accordingly.
    
    Use a kmer size of 55, the same as is used in the paper

```{bash}
~/SPAdes-3.13.0-Linux/bin/spades.py --meta -k 55 -1 ../trimmomatic_cropped/SRR5853122/SRR5853122_trimmed_cropped_1.fastq -2 ../trimmomatic_cropped/SRR5853122/SRR5853122_trimmed_cropped_2.fastq -o ./SRR5853122
```




# Secondary analysis

Use MEGAN6


